functions/Get-DbaSsisEnvironment.ps1

```diff
@@ -0,0 +1,229 @@
+
+dbatools PowerShell module (https://dbatools.io)
+Copyright (C) 2016 Chrissy LeMaire
+This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
+This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
+You should have received a copy of the GNU General Public License along with this program. If not, see http://www.gnu.org/licenses/.
+#>
+function Get-DbaSsisEnvironment {
+
+[CmdletBinding()]
+	Param (
+		[parameter(Mandatory, ValueFromPipeline)]
+		[Alias('SqlServer', 'ServerInstance')]
+		[DbaInstanceParameter[]]$SqlInstance,
+        [Parameter(Mandatory=$false)]
+        [System.Management.Automation.PSCredential]$SqlCredential,
+        [parameter(Mandatory=$false)]
+		[object[]]$Environment,
+        [parameter(Mandatory=$false)]
+        [object[]]$EnvironmentExclude,
+        [parameter(Mandatory=$false)]
+        [object[]]$Folder,
+        [parameter(Mandatory=$false)]
+        [object[]]$FolderExclude,
+		[switch]$Silent
+	)
+
+    begin {
+
+    }
+
+    process {
+
+        foreach($instance in $SqlInstance)
+        {
+            try
+            {
+                Write-Message -Message "Connecting to $instance" -Level 5 -Silent $Silent
+                $connection = Connect-SqlInstance -SqlInstance $instance
+            }
+            catch
+            {
+                Stop-Function -Message "Failed to connect to: $instance" -ErrorRecord $_ -Target $instance -Continue -Silent $Silent
+                return
+            }
+            
+
+            if ($connection.versionMajor -lt 11)
+            {
+                Stop-Function -Message "SSISDB catalog is only available on Sql Server 2012 and above, exiting." -Silent $Silent
+                return
+            }
+
+            try
+            {
+                $ISNamespace = "Microsoft.SqlServer.Management.IntegrationServices"
+
+                Write-Message -Message "Connecting to $instance Integration Services" -Level 5 -Silent $Silent
+                $SSIS = New-Object "$ISNamespace.IntegrationServices" $connection
+            }
+            catch
+            {
+                Stop-Function -Message "Could not connect to Integration Services on $instance" -Silent $Silent
+                return
+            }
+
+            Write-Message -Message "Fetching SSIS Catalog and its folders" -Level 5 -Silent $Silent
+            $catalog = $SSIS.Catalogs | Where-Object { $_.Name -eq "SSISDB" }
+
+            # get all folders names if none provided
+            if($null -eq $Folder) {$Folder = $catalog.Folders.Name}
+
+            # filter unwanted folders
+            if ($FolderExclude) {
```

Comment:
I think the logic on 'Folder', 'FolderExclude', 'Environment' and 'EnvironmentExclude' is much more complex.
To catch all possibilities, it probably needs parameter sets, or we should choose 2 instead of 4 parameters.

If a user enters the same value in -Folder and -FolderExclude -> we will arrive at ```foreach ($f in $folder)```, but $folder will be empty,
If a user enters an environmentname in -Environment and multiple folders in -Folder, no environments will be returned from the folder that does not contain those, while the user may expect to see all environments,...

In short, I think we need to catch an empty collection,
and also explain in the examples and description that those parameters act as additive filters and only environments that pass all filters will be returned.